Initial free threading port

mitsuhiko opened the pull request:
[Triggered by this tweet](https://x.com/etrepum/status/1980144816664252758) I was really curious if Claude can implement free threading support on my bike ride to the office.  The answer seems to be: yes.

This now is a much more complex implementation than the initial commit which only removed the GIL.  This implementation now is makes it both sub-interpreter and free-threading ready.  Given that the code also still supports Python 2.x there are now quite a few paths to take care of.

Important notes: the standard library's JSON module uses a macro called `Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST` and `PySequence_Fast`.  The former uses private APIs which are not exposed and as such we have no option to do that.  The new code special cases lists and tuples and then falls back to the iterator interface.  I'm generally not sure if that is the right approach but this is what some back and forth with codex and claude suggested as the optimal solutions given the constraints.

I'm actually not entirely convinced because I did not run benchmarks.

Unfortunately this change overall drives up the complexity significantly given the many versions that this code needs to support.  For instance in Python 2 I don't think there is a way to avoid the static types so the code also needs to support that.

The code as it is produced right now is ~50% Claude, ~40% Codex and ~10% manual edits and has been going through multiple iterations of Codex based reviews.  I have manually reviewed it a few times now but if you think that approach makes any sense, I would probably make a proper manual pass over still.

mitsuhiko committed:
Initial port

etrepum commented on .github/workflows/build-and-deploy.yml:
```diff
@@ -30,6 +30,21 @@ jobs:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.13t'
```
```suggestion
    name: Tests with free-threading on Python 3.14t
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: '3.14t'
```

mitsuhiko commented:
Free threading just gets rid of the GIL.  The subinterpreter changes move the GIL on a per-subinterpreter level which means that objects can not be shared.  However you are right that these changes do not allow `simplejson` to be loaded into subinterpreters.  That would be a separate change.

etrepum reviewed (commented):
The local changes look mostly ok, especially for decoding, I think the big question here is probably related to iterating objects and arrays during encoding. I think those probably need some kind of locks, but these can be arbitrarily nested too so I don't know if the simple strategies are going to work. Probably worth reading through some of Python's current json module to see what it does.

I think the memo dicts should probably be mostly lock-free but it's probably also possible for people to use those incorrectly with the current API, didn't take a closer look at that but that's why I left those critical sections alone.

mitsuhiko committed:
Added subinterpreter support

This moves shared state onto the module state so that sub-interpreters
work when the GIL is not enabled.

mitsuhiko commented:
Disregard the last commit, I need to do more changes.

mitsuhiko committed:
Migrate to heap types

mitsuhiko committed:
Act on review comments

mitsuhiko committed:
Python 2.x module init

mitsuhiko committed:
Static types for Python 2.x

mitsuhiko committed:
Remove now unused critical section macros

mitsuhiko committed:
Fixes for 2.7

mitsuhiko committed:
Synchronize iteration

mitsuhiko committed:
Lock entire sequence

mitsuhiko commented:
> Probably worth reading through some of Python's current json module to see what it does.

The unfortunate reality is that the stdlib uses `Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST` which is private.  I'm honestly not sure what the correct approach is here.

mitsuhiko committed:
Remove assertions that can fail on shutdown

mitsuhiko committed:
Alternative approach for iteration

etrepum commented:
I think for iteration all that's needed is to essentially copy that macro? AFAICT _PyCriticalSection_Begin is just the static inline internal version of PyCriticalSection_BeginMutex. 

https://github.com/python/cpython/blob/main/Include/internal/pycore_critical_section.h#L29-L42

mitsuhiko committed:
Removed private header and nicer fallback logic

mitsuhiko commented:
@etrepum unfortunately those APIs are private and not part of the stable ABI.  That will only change with 3.15: https://github.com/python/cpython/pull/135899

etrepum commented:
Did this separately in a few claude sessions after some bug fixes came in
